arch: arm: SecureFault Handling for Cortex-M33

ioannisg opened the pull request:
This pull request contributes the implementation of the SecureFault
handling for ARMv8-M-based Cortex-M33. The implementation is
compiled conditionally with compile-time directive
CONFIG_ARM_SECURE_FIRMWARE, which is to signify the intention to
build a Secure image on ARMv8-M with Security Extensions.

This PR implements #6372 for Cortex-M33. 
Consult #6372 for further details.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: SecureFault Handling for Cortex-M33

This commit contributes the implementation of the SecureFault
handling for ARMv8-M-based Cortex-M33. The implementation is
compiled conditionally with compile-time directive
CONFIG_ARM_SECURE_FIRMWARE, which is to signify the intention to
build a Secure image on ARMv8-M with Security Extensions.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

galak commented on arch/arm/core/fault.c:
```diff
@@ -72,8 +72,13 @@ void _FaultDump(const NANO_ESF *esf, int fault)
 				  : "Bus fault on vector table read\n");
 	}
 
+#if !defined(CONFIG_ARMV8_SECURE_FIRMWARE)
```
Rather than an ifdef / else. how about just printing SFSR on its own line.

ioannisg replied:
Sure

galak reviewed (changes requested):
Looks good, where do we define CONFIG_ARM_SECURE_FIRMWARE (not seeing Kconfig anywhere).

one other minor code style comment.

ioannisg commented:
@galak Thanks for your review. Regarding ``CONFIG_ARM_SECURE_FIRMWARE``; the definition of this K-config option is not part of the current PR (but will be handled under a separate issue).

FYI:
I would define ``CONFIG_ARM_SECURE_FIRMWARE`` in arch/arm/core/Kconfig, similarly to the ``ARM_STACK_PROTECTION`` option.
 
``CONFIG_ARM_SECURE_FIRMWARE`` option will reflect the combination of user configuration ("I want to build a Secure image") & hardware capabilities (e.g. "Cortex_M_HAS_Security_Extensions")

ioannisg commented:
PR Updated & rebased.

galak reviewed (changes requested):
Since this change uses ARM_SECURE_FIRMWARE we should also introduce the Kconfig change in this patch

ioannisg committed:
arch: arm: define the ARM_SECURE_FIRMWARE option

This commit introduces the ARM_SECURE_FIRMWARE k-option,
which indicates that we are building an ARM Secure application.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@galak I pushed a commit that defines the option. Let me know what you think.

The overall structure of k-config dependencies for TEE will be addressed in #6374

galak commented:
One thing, not related to this, we need to look at getting cortex-M23/33 support built into the zephyr SDK.

galak reviewed (commented):
Looks good, curious if the name ARM_SECURE_FIRMWARE is coming from some ARM specs / docs or just something you came up with.

agross-oss reviewed (approved):
Looks good.  Nice kconfig description.

ioannisg commented:
@galak Yes, this is correct about sdk. 

Regarding ARM_SECURE_FIRMWARE : I came up with this , inspired by ARM Platform Security Architecture (Overview spec is publicly available) nomenclature. 

The kconfig option text is based on Arm Trustzone and v8 terminology.

Eventually, we will need a CONFIG_ARM_NON_SECURE_FIRMWARE , too, for non secure images on Armv8 mcus with Trustzone...

carlescufi commented on arch/arm/core/Kconfig:
```diff
@@ -45,6 +45,19 @@ config ARM_STACK_PROTECTION
+	bool
+	default n
+	help
+	  This option indicates that we are building a Secure application for
```
I am not sure about using "application" here. Maybe "image" is better? or will we later generate flash images that contain both secure and non-secure code?

ioannisg replied:
@carlescufi IMO, image should refer to the combination of secure and non secure applications that will share the mcu 

That's why I picked the "application"

carlescufi replied:
But will they phisically be an image, or two? I assume a single one, since there will be a single instance of the kernel and libraries?

ioannisg replied:
It will be a single image, but kernel or lib sharing depends on use case, IMO. I don't think we need to address this in this PR.

agross-linaro replied:
it was my understanding that ARM was going to use a modified bootloader and use two separate images, one for secure and one for non-secure.

ioannisg replied:
@carlescufi Changed, "application" to "firmware". Taken directly from the ARM PSA overview documentation.

carlescufi commented on arch/arm/core/Kconfig:
```diff
@@ -45,6 +45,19 @@ config ARM_STACK_PROTECTION
+	default n
+	help
+	  This option indicates that we are building a Secure application for
+	  an ARMv8-M-based MCU, which supports the Security Extension.
```
remove the comma, and also Armv8-M

ioannisg replied:
Comma is ok, as for Arm, we have agreed to keep the style compliant with the earlier ARM architectures.

ioannisg replied:
[Agreed in #4983]

carlescufi commented on arch/arm/core/Kconfig:
```diff
@@ -45,6 +45,19 @@ config ARM_STACK_PROTECTION
+	  A Secure application shall run in Secure state.
+
+	  A Secure application has access to both the Secure and Non-Secure
+	  resources of the Cortex-M MCU. It may generate SecureFaults, if
```
remove the comma

ioannisg replied:
ok

carlescufi reviewed (changes requested):
Looks good, some doc nitpicks since we're entering new territory

ioannisg commented:
Re-based & fixed comments

ioannisg commented:
@carlescufi  Could you have another look at this one?

carlescufi commented on arch/arm/core/Kconfig:
```diff
@@ -42,6 +42,19 @@ config ARM_STACK_PROTECTION
+	bool
+	default n
+	help
+	  This option indicates that we are building a Secure firmware for
```
I am not convinced about using firmware as a countable noun. I don't think one can say "one firmware and two firmwares". I feel we need to clarify this further.

ioannisg commented:
@carlescufi  thanks for your input on this. I updated the PR with the new text for the ARM_SECURE_FIRMWARE.
